crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -2765,6 +2765,183 @@ unsafe impl SystemParam for FilteredResourcesMut<'_, '_> {
     }
 }
 
+/// A [`SystemParam`] that allows running other systems.
+///
+/// To be useful, this must be configured using a [`SystemRunnerBuilder`](crate::system::SystemRunnerBuilder)
+/// to build the system using a [`SystemParamBuilder`](crate::prelude::SystemParamBuilder).
+///
+/// Also see the macros [`compose`](crate::system::compose) and [`compose_with`](crate::system::compose_with)
+/// for some nice syntax on top of this API.
+///
+/// # Examples
+///
+/// ```
+/// # use bevy_ecs::{prelude::*, system::*};
+/// #
+/// # #[derive(Component)]
+/// # struct A;
+/// #
+/// # #[derive(Component)]
+/// # struct B;
+/// # let mut world = World::new();
+/// #
+/// fn count_a(a: Query<&A>) -> u32 {
+///     a.count()
+/// }
+///
+/// fn count_b(b: Query<&B>) -> u32 {
+///     b.count()
+/// }
+///
+/// let get_sum = (
+///     ParamBuilder::system(count_a),
+///     ParamBuilder::system(count_b)
+/// )
+/// .build_state(&mut world)
+/// .build_system(
+///     |mut run_a: SystemRunner<(), u32>, mut run_b: SystemRunner<(), u32>| -> Result<u32, RunSystemError> {
+///         let a = run_a.run()?;
+///         let b = run_b.run()?;
+///         Ok(a + b)
+///     }
+/// );
+/// ```
+pub struct SystemRunner<'w, 's, In = (), Out = (), Sys = dyn System<In = In, Out = Out>>
+where
+    In: SystemInput,
+    Sys: System<In = In, Out = Out> + ?Sized,
+{
+    world: UnsafeWorldCell<'w>,
+    system: &'s mut Sys,
+}
+
+impl<'w, 's, In, Out, Sys> SystemRunner<'w, 's, In, Out, Sys>
+where
+    In: SystemInput,
+    Sys: System<In = In, Out = Out> + ?Sized,
+{
+    /// Run the system with input.
+    #[inline]
+    pub fn run_with(&mut self, input: In::Inner<'_>) -> Result<Out, RunSystemError> {
+        // SAFETY:
+        // - all accesses are properly declared in `init_access`
+        unsafe {
+            self.system
+                .validate_param_unsafe(self.world)
+                .map_err(RunSystemError::Skipped)?;
+        }
+        // SAFETY:
+        // - all accesses are properly declared in `init_access`
+        unsafe { self.system.run_unsafe(input, self.world) }
+    }
+}
+
+impl<'w, 's, Out, Sys> SystemRunner<'w, 's, (), Out, Sys>
+where
+    Sys: System<In = (), Out = Out> + ?Sized,
+{
+    /// Run the system.
+    #[inline]
+    pub fn run(&mut self) -> Result<Out, RunSystemError> {
+        self.run_with(())
+    }
+}
+
+/// The [`SystemParam::State`] for a [`SystemRunner`].
+pub struct SystemRunnerState<Sys: System + ?Sized> {
+    pub(crate) system: Option<Box<Sys>>,
+    pub(crate) access: FilteredAccessSet,
```

Comment:
It's unfortunate that we need to store the `access` like this, since it loses some of the gains from #19477.  ... Ah, I see, we get it when we initialize the inner system, but need to have it available during `init_access`.  

Hmm, would it work to delay initialization of the system until `init_access`?  We require `init_access` to be called before `get_param`, so it would still get done.